src/particle.cpp

```diff
@@ -0,0 +1,225 @@
+{
+  if (n_secondary == MAX_SECONDARY) {
+    fatal_error("Too many secondary particles created.");
+  }
+
+  int64_t n = n_secondary;
+  secondary_bank[n].particle = type;
+  secondary_bank[n].wgt = wgt;
+  std::copy(coord[0].xyz, coord[0].xyz + 3, secondary_bank[n].xyz);
+  std::copy(uvw, uvw + 3, secondary_bank[n].uvw);
+  secondary_bank[n].E = E;
+  if (!run_CE) secondary_bank[n].E = g;
+  n_secondary += 1;
+}
+
+void
+Particle::initialize()
+{
+  // Clear coordinate lists
+  clear();
+
+  // Set particle to neutron that's alive
+  type  = NEUTRON;
+  alive = true;
+
+  // clear attributes
+  surface           = 0;
+  cell_born         = 0;
+  material          = 0;
+  last_material     = 0;
+  last_sqrtkT       = 0;
+  wgt               = 1.0;
+  last_wgt          = 1.0;
+  absorb_wgt        = 0.0;
+  n_bank            = 0;
+  wgt_bank          = 0.0;
+  sqrtkT            = -1.0;
+  n_collision       = 0;
+  fission           = false;
+  delayed_group     = 0;
+  for (int i=0; i<MAX_DELAYED_GROUPS; ++i) {
+    n_delayed_bank[i] = 0;
+  }
+  g = 0;
+
+  // Set up base level coordinates
+  coord[0].universe = C_NONE;
+  n_coord = 1;
+  last_n_coord = 1;
+}
+
+void
+Particle::from_source(const Bank* src, bool run_CE, const double* energy_bin_avg)
+{
+  // set defaults
+  initialize();
+
+  // copy attributes from source bank site
+  type             = src->particle;
+  wgt              = src->wgt;
+  last_wgt         = src->wgt;
+  std::copy(src->xyz, src->xyz + 3, coord[0].xyz);
+  std::copy(src->uvw, src->uvw + 3, coord[0].uvw);
+  std::copy(src->xyz, src->xyz + 3, last_xyz_current);
+  std::copy(src->xyz, src->xyz + 3, last_xyz);
+  std::copy(src->uvw, src->uvw + 3, last_uvw);
+  if (run_CE) {
+    E = src->E;
+    g = 0;
+  } else {
+    g = static_cast<int>(src->E);
+    last_g = static_cast<int>(src->E);
+    E = energy_bin_avg[g - 1];
+  }
+  last_E = E;
+}
+
+void
+Particle::mark_as_lost(const char* message)
+{
+  // Print warning and write lost particle file
+  warning(message);
+  write_restart();
+
+  // Increment number of lost particles
+  alive = false;
+#pragma omp atomic
+  openmc_n_lost_particles += 1;
+
+  // Count the total number of simulated particles (on this processor)
+  auto n = openmc_current_batch * gen_per_batch * openmc_work;
+
+  // Abort the simulation if the maximum number of lost particles has been
+  // reached
+  if (openmc_n_lost_particles >= MAX_LOST_PARTICLES &&
+      openmc_n_lost_particles >= REL_MAX_LOST_PARTICLES*n) {
+    fatal_error("Maximum number of lost particles has been reached.");
+  }
+}
+
+void
+Particle::write_restart()
+{
+  // Dont write another restart file if in particle restart mode
+  if (openmc_run_mode == RUN_MODE_PARTICLE) return;
+
+  // Set up file name
+  // TODO: Add path_output when available on C++ side
+  std::stringstream filename;
+  filename << "particle_" << openmc_current_batch << '_' << id << ".h5";
```

Comment:
I would prefer you resolve this `TODO` now.  If it's too annoying to pull the Fortran character into C++, then just redundantly read it from XML in both C++ and F90.